ios/PriceWidget/Providers/IconProvider.swift

```diff
@@ -0,0 +1,38 @@
+//
+//  IconProvider.swift
+//  Rainbow
+//
+//  Created by Ben Goldberg on 11/8/21.
+//
+
+import Foundation
+import UIKit
+
+@available(iOS 14.0, *)
+final class IconProvider {
+  static let shared = IconProvider()
+  
+  private init() {}
+  
+  public func getIcon(token: String, address: String) -> UIImage? {
+    var icon: UIImage? = UIImage(named: "coinIcons/" + token.lowercased())
+    
+    if (icon != nil) {
+      return icon
+    }
+    
+    let url = URL(string: "https://raw.githubusercontent.com/rainbow-me/assets/master/blockchains/ethereum/assets/\(address)/logo.png")!
+    
+    let semaphore = DispatchSemaphore(value: 0)
```

Comment:
why do we need the semaphore?

Reply:
otherwise the function will return before receiving a response from the data task